katz/src/main/kotlin/katz/instances/Composed.kt

```diff
@@ -0,0 +1,54 @@
+package katz
+
+/**
+ * https://www.youtube.com/watch?v=wvSP5qYiz4Y
+ */
+interface ComposedType<out F, out G>
+
+@Suppress("UNCHECKED_CAST")
+fun <F, G, A> HK<F, HK<G, A>>.lift(): HK<ComposedType<F, G>, A> =
+        this as HK<ComposedType<F, G>, A>
+
+@Suppress("UNCHECKED_CAST")
+fun <F, G, A> HK<ComposedType<F, G>, A>.lower(): HK<F, HK<G, A>> =
+        this as HK<F, HK<G, A>>
+
+open class ComposedFoldable<F, G>(val FF: Foldable<F>, val GF: Foldable<G>) : Foldable<ComposedType<F, G>> {
+    override fun <A, B> foldL(fa: HK<ComposedType<F, G>, A>, b: B, f: (B, A) -> B): B =
+            FF.foldL(fa.lower(), b, { bb, aa -> GF.foldL(aa, bb, f) })
+
+    fun <A, B> foldLC(fa: HK<F, HK<G, A>>, b: B, f: (B, A) -> B): B =
+            foldL(fa.lift(), b, f)
+
+    override fun <A, B> foldR(fa: HK<ComposedType<F, G>, A>, lb: Eval<B>, f: (A, Eval<B>) -> Eval<B>): Eval<B> =
+            FF.foldR(fa.lower(), lb, { laa, lbb -> GF.foldR(laa, lbb, f) })
+
+    fun <A, B> foldRC(fa: HK<F, HK<G, A>>, lb: Eval<B>, f: (A, Eval<B>) -> Eval<B>): Eval<B> =
+            foldR(fa.lift(), lb, f)
+
+    companion object {
+        inline operator fun <reified F, reified G> invoke(FF: Foldable<F> = foldable<F>(), GF: Foldable<G> = foldable<G>()) =
+                ComposedFoldable(FF, GF)
+    }
+}
+
+inline fun <F, reified G> Foldable<F>.compose(GT: Foldable<G> = foldable<G>()) =
```

Comment:
I'm following the practice of breaking to the next line to separate signature and content. You can see the implementation at a glance even for short methods, rather than scrolling horizontally.